### ISSUE TITLE
Colours for directional arrows in forceNetwork

### ISSUE THREAD

**Issue by christophergandrud** [christophergandrud/networkD3#99]
I've been doing a little work on a new branch ([directional](https://github.com/christophergandrud/networkD3/tree/directional)) with directional arrows in `forceNetwork` (an often requested feature https://github.com/christophergandrud/networkD3/issues/71, https://github.com/christophergandrud/networkD3/issues/47, https://github.com/christophergandrud/networkD3/issues/7).

Using the new `arrows` argument in:

``` R
library(networkD3)

data(MisLinks)
data(MisNodes)

forceNetwork(Links = MisLinks, Nodes = MisNodes, Source = 'source',
             Target = 'target', Value = 'value', NodeID = 'name',
             Group = 'group', opacity = 0.8, arrows = T, zoom = T,
             linkDistance = 100)
```

Produces: 

![screen shot 2016-01-09 at 15 35 42](https://cloud.githubusercontent.com/assets/1285805/12216572/b024b028-b6e6-11e5-8124-6a097879022f.png)

The colouring is not currently dynamic and any help fixing this would be great!

Specifically, the arrow colour is now hard coded at `#4679BD`. It would be great to have it follow the link colour which is set by `.style("stroke", function(d) { return d.colour ; })`. However, replacing this for the [current call](https://github.com/christophergandrud/networkD3/blob/directional/inst/htmlwidgets/forceNetwork.js#L257) `.style("stroke", "#4679BD")` simply results in no arrows being created.

**Comment by cjyetman** [comment]
I have some ideas about how to do this. Was following the link color the only thing holding this back, or were there other considerations?

**Comment by cjyetman** [comment]
I have a branch where this is working, including matching the link colour, here: https://github.com/cjyetman/networkD3/tree/directed_arrows

to test, one can install with...
```r
devtools::install_github('cjyetman/networkD3', ref = 'directed_arrows')
```

The more difficult problem here is positioning the arrows correctly. Does anyone have thoughts or ideas about that? The reason it is difficult is because the end of each link is at the center of each node, so the arrows need to be pushed back along the link line to not be obscured by the node. But a combination of potential options, like node opacity, node size/radius,  and link weight/stroke-width, make determining exactly how far from the end of the line the arrow should start rather complicated.

I'm tempted to suggest that maybe the link lines themselves should end at the edge of the node circles rather than at the center point, which looks better when the node circles are transparent... and that would also make the arrow placement much easier. Is there any reason to not go down that route?

**Comment by cjyetman** [comment]
the following code produces the following plot (when using my`directed_arrows` branch version)...
```r
library(networkD3)

URL <- paste0("https://cdn.rawgit.com/christophergandrud/networkD3/",
              "master/JSONdata/miserables.json")
MisJson <- jsonlite::fromJSON(URL)
ValjeanInds <- which(MisJson$links == 11, arr = TRUE)[, 1]
ValjeanCols <- ifelse(1:nrow(MisJson$links) %in% ValjeanInds, "#bf3eff", "#666")

forceNetwork(Links = MisLinks, Nodes = MisNodes, Source = "source", Target = "target", 
             Value = "value", NodeID = "name", Group = "group", opacity = 0.8, 
             linkColour = ValjeanCols, arrows = TRUE, zoom = TRUE)
```
<img width="536" alt="mis_colour" src="https://cloud.githubusercontent.com/assets/1708872/22970316/ff8bc7fa-f371-11e6-85aa-bf7e58c0b1c3.png">

**Comment by christophergandrud** [comment]
Thanks for this @cjyetman. 

If I remember correctly the arrow placement was the other hold up on this. Yep, I guess the ideal would be that the links end at the node edge. Did you have a mock up of this?

**Comment by cjyetman** [comment]
I don't have a mock up, but I imagine it's possible, in the `tick` function, to calculate a line that offsets itself from `(source.x, source.y)` and `(target.x, target.y)` by the radius of the node circle. I'll play with that idea when I get a chance.

**Comment by cjyetman** [comment]
My 'directed_arrows' branch has been updated to end the link lines at the edges of the node circles. It's ugly code, and ugly math, but it's a functional mock up.

install the dev branch with...
```r
devtools::install_github('cjyetman/networkD3', ref = 'directed_arrows')
```

this code...
```r
URL <- paste0("https://cdn.rawgit.com/christophergandrud/networkD3/",
              "master/JSONdata/miserables.json")
MisJson <- jsonlite::fromJSON(URL)
ValjeanInds <- which(MisJson$links == 11, arr = TRUE)[, 1]
ValjeanCols <- ifelse(1:nrow(MisJson$links) %in% ValjeanInds, "#bf3eff", "#666")

forceNetwork(Links = MisLinks, Nodes = MisNodes, Source = "source", Target = "target", 
             Value = "value", NodeID = "name", Group = "group", opacity = 0.8, 
             linkColour = ValjeanCols, arrows = TRUE, zoom = TRUE)
```
results in...
<img width="521" src="https://cloud.githubusercontent.com/assets/1708872/23000586/7c465106-f3e1-11e6-84bd-5eb988f0acca.png">

**Comment by christophergandrud** [comment]
That looks great!

We should merge this in. 

The ultimate would be if all of the links changed to the contrast colour on mouse over. Though having it be hard coded like this may also be useful for highlighting specific relationships.

**Comment by cjyetman** [comment]
Can you restate that please? I'm not exactly sure what you mean.

mouse-over a node, link, or both?
all links, or all links related to the thing that's being moused-over?

I can imagine something where when you hover over a node, all links and nodes that are **not** related to that node could have their opacity set 50% lower than what it currently is, or something like that?

**Comment by christophergandrud** [comment]
> I can imagine something where when you hover over a node, all links and nodes that are not related to that node could have their opacity set 50% lower than what it currently is, or something like that?

Exactly!

**Comment by cjyetman** [comment]
I have to make 2 different blocks of code for calculating the lines either way because the arrow-heads have to be taken into account when they're being displayed... would you prefer if no arrow-heads are displayed that the exact behavior as before happened (lines go all the way to the center point and can be seen underneath the node if it is partially transparent), or would it be better to also stop the lines without arrow-heads just at the edge of the nodes. I kinda prefer the latter, but that's not really sticking to the backwards-compatible philosophy.

**Comment by christophergandrud** [comment]
I agree that the latter is also preferable. If I understand correctly, you're saying that the non-backwards compatible thing will be that the links stop at the edge of the nodes? This is a fairly minor aesthetic change (that based on the work you did before I kind of prefer). So I would say it's not much of a problem.

**Comment by cjyetman** [comment]
Yeah, it won't really break anything, it just deviates from the previous aesthetic.

**Comment by cjyetman** [comment]
resolved by #182... closing
